repository/src/lib.rs

```diff
@@ -267,10 +268,71 @@ impl<T: RawRepository> DistributedRepository<T> {
     /// and update the corresponding `a-#` branch to it
     pub async fn approve(
         &mut self,
-        _agenda_commit_hash: &CommitHash,
-        _proof: Vec<TypedSignature<Agenda>>,
+        agenda_commit_hash: &CommitHash,
+        proof: Vec<TypedSignature<Agenda>>,
     ) -> Result<CommitHash, Error> {
-        unimplemented!()
+        // Check if the agenda branch is rebased on top of the `finalized` branch.
+        let last_header_commit = self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        let agenda_branch_name = self.raw.get_branches(*agenda_commit_hash).await?;
+        let agenda_branch_name = agenda_branch_name.get(0).unwrap();
+        if self
+            .raw
+            .find_merge_base(last_header_commit, *agenda_commit_hash)
+            .await?
+            != last_header_commit
+        {
+            return Err(anyhow!(
+                "branch {} should be rebased on {}",
+                agenda_branch_name,
+                FINALIZED_BRANCH_NAME
+            ));
+        }
+
+        // Verify agenda with agenda proof
+        let agenda_commit = self.read_commit(*agenda_commit_hash).await?;
+        let agenda = match agenda_commit.clone() {
+            Commit::Agenda(agenda) => agenda,
+            _ => panic!("given commit hash is not agenda commit hash"),
+        };
+        for typed_signature in proof.clone() {
+            typed_signature.verify(&agenda)?;
+        }
```

Comment:
This is not enough. You must use CSV, starting from the finalized block.

Reply:
fixed. I added csv before this.